Java: Add support for Java 21 language features

smowton opened the pull request:
Principally this adds support for Java 21 binding and record patterns. These can occur in both `instanceof` and `case` contexts, and look like `case String s when s.length() > 5` (a binding pattern with a guard), or `case OuterRecord(String field, InnerRecord(int otherField))` (a record pattern with no guard).

Control-flow-wise, these are modelled as type-testing nodes that on success branch to initialise their binding variable declarations. For the data-flow library they are represented by a stack of read steps corresponding to each nested record pattern.

"Traditional" switches with all constant cases continue to have simple fan-out CFGs, where the switch branches to all possible cases in parallel. Pattern switches on the other hand have a CFG more like a series of `if` statements -- so `switch(...) { case String s: ... case Integer i: ... }` gets a similar CFG to `if (e instanceof String s) { ... } else if (e instanceof Integer i) { ... }`.

The most complex case is a block with interleaved constant and pattern cases -- the block `switch(e) { case 1: ... case 2: ... case Integer i | i < 10: ... case 15: ... case 16: ... default: ... }` gets a CFG that resembles `switch(e) { case 1: ... case 2: ... default: if (e instanceof Integer i && i < 10) { ... } else { switch(e) { case 15: ... } } }`

The dbscheme is modified in three ways: 

1. Record destruction requires paying attention to the canonical constructor's field order, so `isCanonicalConstr` is introduced to tag the canonical constructor.
2. The switch case `case null, default:` requires a special tag table `isCaseNullDefault` to distinguish it from `case null:` and `default:`, because our database scheme doesn't record an expression for the `default` label.
3. A new expression kind is introduced for record patterns.

Finally, the existing hackish representation of binding instanceof (`e instanceof T bindingvar`), which represented it like the pseudo-expression `(T bindingvar = e) instanceof T`, is rewritten to store the binding variable explicitly as its own subexpression, thereby making way for record patterns.

smowton commented:
@aschackmull @yo-h @igfoo would you mind casting an eye over this one?

smowton commented:
Snag found, fix pending: not all the little ancillary dataflows (ObjFlow, TypeFlow, DispatchFlow... are there others?) covered both pattern-case and binding-instanceof (rearchitected) as flow steps. Fix coming up.

smowton commented:
Notes from self-review:

1. Obvious followup PR: create a class `SwitchBlock` that factors out the common elements of SwitchStmt and SwitchExpr to reduce code duplication
2. Potential followup PR: factor the common elements of the various local-flow-step relations in dataflow, typeflow, dispatchflow and objflow.
3. Crucial decision: this PR replaces the hackish representation of `x instanceof T v` as if it were a local declaration `T v = x` with a more straightforward AST representation with QL drawing the line between `x` and `v` in the same way it does other dataflow edges. Pro: this allows a clean and uniform handling of the new record-pattern feature `x instanceof TRec(T1 field1, T2 field2, ...)`, as well as pattern cases `case T v` that weren't amenable to the same pseudo-initialiser trick. Con: this means binding-instanceof needed handling in the local-step relations for objflow, typeflow, dispatchflow; there may be more that I haven't noticed (although if so they don't seem to be causing alert differences in QA).

aschackmull commented on java/ql/lib/semmle/code/java/Expr.qll:
```diff
@@ -1531,6 +1546,9 @@ class SwitchExpr extends Expr, StmtParent, @switchexpr {
   }
 
+  /** Holds if this switch has a case handling a null literal. */
+  predicate hasNullCase() { this.getAConstCase().getValue(_) instanceof NullLiteral }
```
`ConstCase`s currently explicitly excludes `case null, default`, so this predicate seems incorrect (same thing goes for the mirrored predicate in `SwitchStmt`).

aschackmull commented on java/ql/lib/semmle/code/java/Expr.qll:
```diff
@@ -1612,18 +1648,46 @@ class LocalVariableDeclExpr extends Expr, @localvariabledeclexpr {
+   */
+  InstanceOfExpr getAssociatedInstanceOfExpr() { result.getPattern().getAChildExpr*() = this }
+
+  /** Holds f this is a declaration stemming from a pattern instanceof expression. */
```
```suggestion
  /** Holds if this is a declaration stemming from a pattern instanceof expression. */
```

aschackmull commented on java/ql/lib/semmle/code/java/Statement.qll:
```diff
@@ -483,7 +527,58 @@ class ConstCase extends SwitchCase {
+ *
+ * Note binding patterns are represented as `LocalVariableDeclExpr`s.
+ */
+class PatternExpr extends Expr {
```
I'd expect this class in `Expr.qll` instead of `Statement.qll`.

aschackmull commented on java/ql/lib/semmle/code/java/Expr.qll:
```diff
@@ -1612,18 +1648,46 @@ class LocalVariableDeclExpr extends Expr, @localvariabledeclexpr {
+  predicate hasAssociatedInstanceOfExpr() { exists(this.getAssociatedInstanceOfExpr()) }
+
   /** Gets the initializer expression of this local variable declaration expression, if any. */
   Expr getInit() { result.isNthChildOf(this, 0) }
```
Probably add another parallel helper to include `getInit` plus the directly assigned exprs for simple patterns.

aschackmull replied:
And use that in `VariableAssign`.

aschackmull commented on java/ql/lib/semmle/code/java/PrettyPrintAst.qll:
```diff
@@ -385,7 +385,12 @@ private class PpInstanceOfExpr extends PpAst, InstanceOfExpr {
+    (
+      result = this.getPattern().asBindingPattern().getName()
+      or
+      result = this.getPattern().asRecordPattern().toString()
```
This case doesn't appear to recursively print the entire record pattern, which it really should. The idea of this library is to avoid intermediate string concatenation, so record patterns should be printed using `getChild` and patterns should have their own `PpAst` extension. Let me know if you need more details on this library - it's very useful for inspecting the output of e.g. extractor desugarings.

aschackmull commented on java/ql/lib/semmle/code/java/PrettyPrintAst.qll:
```diff
@@ -757,8 +764,15 @@ private class PpSwitchCase extends PpAst, SwitchCase {
+          (max(int j | j = 0 or exists(this.(ConstCase).getValue(j))) + this.getCaseDefaultOffset())
   }
 
   override PpAst getChild(int i) {
```
This is missing handling for patterns.

aschackmull commented on java/ql/lib/semmle/code/java/Statement.qll:
```diff
@@ -483,7 +527,58 @@ class ConstCase extends SwitchCase {
+    (
+      this.getParent() instanceof SwitchCase or
+      this.getParent() instanceof InstanceOfExpr or
+      this.getParent() instanceof PatternExpr
```
Couldn't we change this to `RecordPatternExpr` without changing semantics, but thereby getting rid of recursion?
```suggestion
      this.getParent() instanceof RecordPatternExpr
```

aschackmull commented on java/ql/lib/semmle/code/java/dataflow/TypeFlow.qll:
```diff
@@ -116,6 +116,22 @@ private predicate step(TypeFlowNode n1, TypeFlowNode n2) {
   n2.asSsa().(BaseSsaUpdate).getDefiningExpr().(VariableAssign).getSource() = n1.asExpr()
   or
   n2.asSsa().(BaseSsaImplicitInit).captures(n1.asSsa())
+  or
+  exists(PatternCase pc, LocalVariableDeclExpr patternVar |
+    patternVar = pc.getPattern().asBindingPattern() and
+    n2.asSsa().(BaseSsaUpdate).getDefiningExpr() = patternVar and
+    (
+      pc.getSwitch().getExpr() = n1.asExpr()
+      or
+      pc.getSwitchExpr().getExpr() = n1.asExpr()
+    )
+  )
+  or
+  exists(InstanceOfExpr ioe, LocalVariableDeclExpr patternVar |
+    patternVar = ioe.getPattern().asBindingPattern() and
+    n2.asSsa().(BaseSsaUpdate).getDefiningExpr() = patternVar and
+    ioe.getExpr() = n1.asExpr()
+  )
```
Fold these two case into `VariableAssign.getSource()`. Also, the disjunct in the switch case can be simplified using `SwitchCase.getSelectorExpr`.

aschackmull commented on java/ql/lib/semmle/code/java/dataflow/TypeFlow.qll:
```diff
@@ -413,30 +429,41 @@ private predicate downcastSuccessor(VarAccess va, RefType t) {
-  exists(InstanceOfExpr ioe, BaseSsaVariable v |
-    ioe.getExpr() = v.getAUse() and
+private predicate instanceOfGuarded(Expr e, RefType t) {
+  exists(InstanceOfExpr ioe |
```
I would like to see the distinction between the `instanceOfGuarded` predicate here and the `patternCaseGuarded` predicate below refactored into one, by adding an `isInstanceofCheck` member predicate on `Guard`.

aschackmull commented on java/ql/lib/semmle/code/java/dataflow/TypeFlow.qll:
```diff
@@ -413,30 +429,41 @@ private predicate downcastSuccessor(VarAccess va, RefType t) {
-    guardControls_v1(ioe, aa.getBasicBlock(), true)
+private predicate patternCaseGuarded(Expr e, RefType t) {
+  exists(PatternCase pc |
+    e = getAProbableAlias([pc.getSwitch().getExpr(), pc.getSwitchExpr().getExpr()]) and
```
The pattern `[pc.getSwitch().getExpr(), pc.getSwitchExpr().getExpr()]` already has an existing convenience method `getSelectorExpr`.

aschackmull commented on java/ql/lib/semmle/code/java/dataflow/TypeFlow.qll:
```diff
@@ -413,30 +429,41 @@ private predicate downcastSuccessor(VarAccess va, RefType t) {
   )
 }
 
+private Expr getAProbableAlias(Expr e) {
```
This refactor is inherently quadratic, so this looks dangerous. The proper fix here is to add (maybe just a tiny bit of) value numbering. It could still be private and only cover these two cases, but it should be a def-use relation rather than a use-use cartesian product.

aschackmull commented on java/ql/lib/semmle/code/java/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -234,6 +234,17 @@ predicate storeStep(Node node1, ContentSet f, Node node2) {
+private Field getLexicallyOrderedRecordField(Record r, int idx) {
+  result =
+    rank[idx + 1](Field f, int i, Parameter p |
+      f = r.getAField() and
+      p = r.getCanonicalConstructor().getParameter(i) and
+      f.getName() = p.getName()
```
I guess this is unlikely to be a performance problem any time soon, but as-is it's guaranteed to be suboptimal, since there's no efficient join-ordering as written.

aschackmull commented on java/ql/lib/semmle/code/java/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -190,14 +190,21 @@ predicate simpleAstFlowStep(Expr e1, Expr e2) {
   e2 = any(NotNullExpr nne | e1 = nne.getExpr())
   or
   e2.(WhenExpr).getBranch(_).getAResult() = e1
+  or
+  exists(SwitchExpr se | e1 = se.getExpr() and e2 = se.getACase().(PatternCase).getPattern())
+  or
+  exists(SwitchStmt ss | e1 = ss.getExpr() and e2 = ss.getACase().(PatternCase).getPattern())
+  or
+  exists(InstanceOfExpr ioe | e1 = ioe.getExpr() and e2 = ioe.getPattern())
```
I think we generally want the flow for simple patterns to be included in `VariableAssign`. That means that the `e1` expression here steps directly through the ssa definition to the first use via `simpleLocaLFlowStep0` below, so we should avoid creating parallel paths in those cases and thus exclude simple patterns in these three disjuncts.
Adding support in `VariableAssign.getSource` will also cover the currently missing case in `closureFlowStep` in `DataFlowPrivate.qll`.

aschackmull replied:
There's another missing flow step in `localMustFlowStep`, which will also be handled once `VariableAssign.getSource` covers simple patterns.

smowton replied:
I've amended these disjuncts to only care about record patterns, where we need an intermediate node to terminate the read step (or messily use ssa-step in defining the read step).

For example, simple pattern `x instanceof T v ... use(v)` now looks like `x -flow-> v_use`, but `x instanceof R1(R2(t v)) ... use(v)` looks like `x -flow-> R1 -read-> R2 -read-> v -flow-> v_use`.

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/UnreachableBlocks.qll:
```diff
@@ -170,10 +170,11 @@ class ConstSwitchStmt extends SwitchStmt {
   /** Gets the matching case, if it can be deduced. */
   SwitchCase getMatchingCase() {
     // Must be a value we can deduce
+    // TODO: handle other known constants (enum constants, String constants)
```
Probably doesn't matter - I think this code is dead.

aschackmull commented on java/ql/lib/upgrades/ecfcf050952e54b1155fc89525db84af6ad34aaf/exprs.ql:
```diff
@@ -0,0 +1,47 @@
+  then (
+    newParent = getParent(getParent(e)) and newIndex = 0
+  ) else (
+    // Initialiser moves to hang directly off the instanceof expression
```
Looks like this and the next comment dropped a few lines down (likely due to autoformat). We should move them up a bit to make this less confusing.

aschackmull commented on java/ql/src/Likely Bugs/Collections/ReadOnlyContainer.ql:
```diff
@@ -40,6 +40,6 @@ where
-  // and `v` is not implicitly initialized by a for-each loop.
-  not exists(EnhancedForStmt efs | efs.getVariable().getVariable() = v)
+  // and `v` is not implicitly initialized.
+  not v.(LocalVariableDecl).getDeclExpr().hasImplicitInit()
```
`v.getAnAssignedValue()` should also account for simple patterns. This makes me more and more likely to  want to simply include the simple patterns in `LocalVariableDeclExpr.getInit` and hence exclude them from `hasImplicitInit`.

smowton replied:
These are now split so that simple patterns are direct and record patterns are implicit. Note that many of these were already wrong in that they didn't account for catch variables (though that is of course less likely to arise in practice)

aschackmull commented on java/ql/test/library-tests/object-tostring-flow-binding-patterns/Test.java:
```diff
@@ -0,0 +1,29 @@
+
+  public static String test(Object o) {
+
+    if (o instanceof Object o2) {
```
Should this have been s/`Object`/`Intf`/?

aschackmull replied:
Ah wait, I guess not, as that would probably count as being a too concrete type and hence not make the ObjectFlow lib kick in.

aschackmull commented on java/ql/test/library-tests/dataflow/collections/options:
```diff
@@ -0,0 +1 @@
+//semmle-extractor-options: --javac-args --release 21
```
There are no other changes in this dir? What does this change accomplish?

smowton replied:
Removed

aschackmull commented on java/ql/test/query-tests/Nullness/G.java:
```diff
@@ -0,0 +1,22 @@
+      default -> System.out.println("Something else");
+    }
+
+    switch(s) { // BAD; lack of a null case means this may throw.
```
Let's add a test here for a switch with `case null, default`. Maybe make it a switch expression just to change things around a bit.

igfoo commented on java/ql/lib/change-notes/2023-11-03-jdk21-support.md:
```diff
@@ -0,0 +1,5 @@
+---
+category: minorAnalysis
+---
+* Switch cases using patterns and both cases and instanceof expressions using record patterns are now supported. The new class `RecordPatternExpr` is introduced to represent patterns, `PatternCase` and `CaseNullDefault` to represent new kinds of case statement, `InstanceOfExpr` gains `getPattern` to replace `getLocalVariableDeclExpr`.
```
`Switch cases using patterns and both cases ...` is a little confusing to read. I think splitting this into 2 bullet points might be better.

igfoo commented on java/downgrades/dee651b58d1e5455ca2d07eca37775a21d772fcc/exprs.ql:
```diff
@@ -0,0 +1,55 @@
+  (
+    if oldKind = 89
+    then /* record pattern */ newKind = 74
+    else /* error expression */ oldKind = newKind
```
This is not just error expressions, but also all the other kinds of expressions, right?

smowton replied:
Yes. Autoformat strikes again.

aschackmull commented:
Could we add a test that checks whether default cases (both the regular and the `null, default` case) act as negated instanceof checks when there are unguarded type check cases (including both simple type patterns and possibly also record patterns (which would require a notion of always-matching for nested patterns in order to conclude that the outer type was the failure reason)).
The virtual dispatch library makes use of negated type checks to filter otherwise potential call targets, cf. `VirtualDispatch.qll::impossibleDispatchTarget`.

aschackmull commented on java/ql/lib/semmle/code/java/dispatch/ObjFlow.qll:
```diff
@@ -80,6 +80,20 @@ private predicate step(Node n1, Node n2) {
       for.getExpr() = n1.asExpr()
     )
+    or
+    exists(PatternCase pc |
```
These couple of added cases should be superfluous once they are included in `VariableAssign.getSource()`.

aschackmull commented on java/ql/lib/semmle/code/java/dispatch/DispatchFlow.qll:
```diff
@@ -167,6 +167,20 @@ private module TypeTrackingSteps {
+        (
+          pc.getSwitch().getExpr() = n1.asExpr()
+          or
+          pc.getSwitchExpr().getExpr() = n1.asExpr()
```
Same here: These couple of added cases should be superfluous once they are included in `VariableAssign.getSource()`.

smowton commented on java/ql/src/Likely Bugs/Statements/MissingEnumInSwitch.ql:
```diff
@@ -17,6 +17,6 @@ from SwitchStmt switch, EnumType enum, EnumConstant missing
   switch.getExpr().getType() = enum and
   missing.getDeclaringType() = enum and
-  not exists(switch.getDefaultCase()) and
+  not exists(switch.getDefaultOrNullDefaultCase()) and
```
Moderately disagree; I think existing QL will be surprised to find a DefaultCase, getDefaultCase etc that have a const operand, match for situations other than default, etc.

aschackmull replied:
I'll try to elaborate my view and thinking here: If there's a `case null, default` then that's very similar to a regular `default` case - in fact if Java allowed a hypothetical ad-hoc pragma annotating the entire switch saying "don't throw on null" and this was applied then the two default cases would be exactly identical. So changing a `default` case to a `case null, default` is just a way to essentially add such a pragma - it's still going to be the case that matches when all the others don't and no other case is going to match null either way.
Also, looking at the query changes seen here, I think that any existing QL that cares about the existence of a default case or want to inspect its body or anything like that would want to treat `case null, default` the same way. It doesn't really matter that it matches null - that's just yet another value that isn't matched by the other cases.

aschackmull replied:
Also, I think it matches the current qldoc for `getDefaultCase` - when it exists, a `case null, default` is "_the_ default case" of the switch.

smowton replied:
Alright, done; getDefaultCase now gets either one and DefaultCase includes NullDefaultCase.

smowton commented:
@aschackmull addressed comments so far, and added negative type-test logic at https://github.com/github/codeql/pull/14671/commits/ede87083dfeb3059ea88df46bc65b5e5f1318b69 (that will surely have further consequences that require review!)

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -175,14 +174,45 @@ class Guard extends ExprParent {
+  }
+
+  /** Holds if this guard tests whether `e` has type `t`. */
+  predicate appliesTypeTest(Expr e, Type t) { e = testedExpr and t = testedType }
```
This is not entirely correct for record patterns, as such patterns might correspond to an implicit conjunction with a sequence of nested type tests. As such, record patterns are always type tests on the true-edge, but only sometimes negative type tests on the false-edge, where the latter depends on the record pattern not containing nested type test that are capable of failing.

smowton replied:
Good point -- new test cases added and `appliesTypeTest` modified to account for that.

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -108,8 +108,7 @@ class Guard extends ExprParent {
     result = this.(Expr).getBasicBlock() or
-    result = this.(SwitchCase).getSwitch().getExpr().getBasicBlock() or
-    result = this.(SwitchCase).getSwitchExpr().getExpr().getBasicBlock()
+    result = this.(SwitchCase).getSelectorExpr().getBasicBlock()
```
This isn't quite what we want for pattern cases (and will also be wrong if we switch the cfg to sequential testing of const cases). This needs to return the basic block that has the outgoing true- and false-edges. I do realise that the existing code only approximated this for const cases if the selector expression wasn't a post-order cfg node, but for pattern cases we can do better.

aschackmull replied:
Also, the `hasBranchEdge` predicate below is wrong for pattern cases. But again, if we switch to a sequential switch cfg construction then it will need even further updates.

smowton replied:
I'll fix `getBasicBlock`, but can't expose this in a test at the moment because all uses of getBasicBlock either require the guard to be an equality test, or else are used to ask dominance questions (in either direction) where I can't see a way to expose the difference due to the top of the switch dominating the case, and the flow of alternate pattern cases only converging outside the switch block, because it is illegal to fall through either into or out of a pattern case.

smowton replied:
I've restricted `hasBranchEdge` to non-pattern cases, because pattern cases constitute ConditionBlocks now -- and if we make the switch-is-linear-series-of-tests change, const cases will be too.

smowton replied:
Looks like the only users where the guard could be a switch case are modulus and sign analysis;  I haven't immediately been able to cook a test to expose this logic.

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -175,14 +174,45 @@ class Guard extends ExprParent {
+  predicate appliesTypeTest(Expr e, Type t) { e = testedExpr and t = testedType }
+}
+
 private predicate switchCaseControls(SwitchCase sc, BasicBlock bb) {
```
I don't think we need to account for `PatternCase` here - rather I think it should be excluded. It ought to be handled by the regular `ConditionBlock.controls` disjunct in the `directlyControls` predicate above.

smowton replied:
Agreed, this is redundant now pattern cases are recognised as condition blocks. Removed. I have however adapted this to note that the predecessor might not be the top of the switch block any longer -- it could be a previous pattern case (and added tests for this).

smowton commented:
@aschackmull this is once again up to date with your comments

smowton commented:
Pushed a minor consistency fix https://github.com/github/codeql/pull/14671/commits/b4afa7f5a4469eaa9cb020d59899e1df7a91c1dd to treat `case null, default` the same as `case null: default:` from a CFG perspective.

This caused a minor snag in identifying unreachable callables -- https://github.com/github/codeql/pull/14671/commits/95d203fe8a4d85666b9137455a19ecf49670cf09 -- but I think this doesn't matter because flowing `case null` and `default` together and then making calls against the switch selector expression is a somewhat odd pattern requiring at least another null guard to be safe.

We could address this by using a slightly less faithful CFG where the null case is only reached after the last pattern case fails, but this strikes me as dubious since (a) it is less faithful to what the JVM will do (null test up front, then start checking the other cases), and (b) it would entail lexically-backwards flow in a switch block, which seems like a can of worms worth avoiding.

smowton commented:
Did some self-review today by looking at users of VariableAssign and CastNode and cousins. Added a commit (the latest) to classify binding nodes as casting nodes and therefore warranting a type test.

This is a bit ugly because we're noting the use as the casting node. Options:

1. Keep it as is and put up with more CastNodes (and possibly more than one per actual syntactic cast / runtime type test)
2. Make the LocalVariableDeclExpr itself the "source" of the assignment and introduce an expr-expr step pointing to it, so the flow goes switch/instanceof-expr -> var-decl-expr -> first use, unlike the situation with a normal local variable declaration where the flow goes straight from the initialiser to its first use. This means less CastNodes, but a VariableAssign with a surprising source (the variable itself, or `VariableAssign::this`), which may well surprise some users.

aschackmull commented on java/ql/test/library-tests/flow-through-binding/Test.java:
```diff
@@ -0,0 +1,42 @@
+  public static Object testFlowThroughSwitchStmt(String s, Integer i, boolean unknown) {
+    Object o = unknown ? s : i;
+    switch (o) {
+      case Integer i2 -> { return i2; }
```
`ReturnNode`s are always in the path, so to make sure we're actually testing that the cast is properly detected, we should cast back to `Object` asap.
```suggestion
      case Integer i2 -> { return (Object)i2; }
```

aschackmull commented on java/ql/test/library-tests/flow-through-binding/Test.java:
```diff
@@ -0,0 +1,42 @@
+
+  public static Object testFlowThroughSwitchExpr(String s, Integer i, boolean unknown) {
+    Object o = unknown ? s : i;
+    Integer toRet = switch (o) {
+      case Integer i2 -> i2;
```
```suggestion
    Object toRet = switch (o) {
      case Integer i2 -> (Object)i2;
```

aschackmull commented:
> This is a bit ugly because we're noting the use as the casting node. Options:
> 
> 1. Keep it as is and put up with more CastNodes (and possibly more than one per actual syntactic cast / runtime type test)
> 2. Make the LocalVariableDeclExpr itself the "source" of the assignment and introduce an expr-expr step pointing to it, so the flow goes switch/instanceof-expr -> var-decl-expr -> first use, unlike the situation with a normal local variable declaration where the flow goes straight from the initialiser to its first use. This means less CastNodes, but a VariableAssign with a surprising source (the variable itself, or `VariableAssign::this`), which may well surprise some users.

Regardless, I think a change here is fine to consider as potential followup work.

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/internal/GuardsLogic.qll:
```diff
@@ -55,9 +55,13 @@ predicate implies_v1(Guard g1, boolean b1, Guard g2, boolean b2) {
   )
   or
-  g1.(DefaultCase).getSwitch().getAConstCase() = g2 and b1 = true and b2 = false
+  exists(DefaultCase sc | g1 = sc |
```
Might as well revert this back to the inline cast it was.

aschackmull commented:
The `getInit` in `Variable.getAnAssignedValue` should be updated to `getInitOrPatternSource()`.

aschackmull commented on java/ql/lib/semmle/code/java/Expr.qll:
```diff
@@ -1540,27 +1562,43 @@ class SwitchExpr extends Expr, StmtParent, @switchexpr {
+   *
+   * For a match against a record pattern, this is the type of the outermost record type.
+   */
+  RefType getCheckedType() {
```
I think the way that this predicate is typically used means that we can't allow it to hold for restricted record patterns, i.e. those for which `isUnrestricted()` doesn't hold.

smowton replied:
Done, though most users seemed interested in the positive case (instanceof passes ==> e is of type) not the negative, so most have been updated to notice even restricted record type checks.

aschackmull replied:
Hmm, right, I see. In any case, I think this is the safer option.

aschackmull commented on java/ql/lib/semmle/code/java/Statement.qll:
```diff
@@ -428,6 +450,20 @@ class SwitchCase extends Stmt, @case {
+  /**
+   * Gets the `i`th case in this case's switch block.
+   */
+  SwitchCase getSiblingCase(int i) {
```
Are we sure that we really want this as a general-purpose convenience predicate? It's inherently quadratic, so we may very well end up in a situation where we're forced to deprecate it for performance reasons.

smowton replied:
Suggest I follow up immediately after this PR to introduce `class SwitchBlock` and thereby remove everything shared between SwitchStmt and SwitchExpr.

aschackmull commented on java/ql/lib/semmle/code/java/Statement.qll:
```diff
@@ -428,6 +450,20 @@ class SwitchCase extends Stmt, @case {
+   * Gets this case's ordinal in its switch block.
+   */
+  int getCaseIndex() {
+    this = this.getSwitch().getCase(result) or this = this.getSwitchExpr().getCase(result)
```
There's actually a superfluous join here - not that I think it matters much, but we could simplify a bit:
```suggestion
    this = any(SwitchStmt s).getCase(result) or this = any(SwitchExpr s).getCase(result)
```

aschackmull commented on java/ql/lib/semmle/code/java/Statement.qll:
```diff
@@ -483,9 +539,36 @@ class ConstCase extends SwitchCase {
+  /** Gets the guard applicable to this pattern case, if any. */
+  Expr getGuard() { result.isNthChildOf(this, -3) }
+
+  override string pp() { result = "case T t ..." }
+
+  override string toString() { result = "case T t ..." }
```
The placeholder names `T` and `t` have been rubbing me the wrong way. How about `<Pattern>` instead? We have a bit of precedence for that sort of toString in `ExprStmt`.
```suggestion
  override string pp() { result = "case <Pattern>" }

  override string toString() { result = "case <Pattern>" }
```

aschackmull commented on java/ql/lib/semmle/code/java/PrettyPrintAst.qll:
```diff
@@ -743,10 +746,14 @@ private class PpSwitchStmt extends PpAst, SwitchStmt {
+
   override string getPart(int i) {
     i = 0 and result = "default" and this instanceof DefaultCase
     or
-    i = 0 and result = "case " and this instanceof ConstCase
+    i = 0 and result = "case " and not this instanceof DefaultCase
```
This needs to be updated now that `NullDefaultCase` is included in `DefaultCase`.

aschackmull replied:
I guess we ought to have a qltest for this library...

aschackmull commented on java/ql/lib/semmle/code/java/PrettyPrintAst.qll:
```diff
@@ -770,6 +784,36 @@ private class PpSwitchCase extends PpAst, SwitchCase {
+    or
+    i = 1 and result = this.getPattern().asRecordPattern()
+    or
+    i = 4 and result = this.getRuleExpression()
```
This index is too low - it clashes with the `" -> "` in `getPart` when this is a simple pattern.

aschackmull replied:
I guess both this line and the disjunct below should simply be `i=5`.

aschackmull replied:
And then we might as well make the `":"` and the `" -> "` be at index 4 unconditionally and delete the `getPatternOffset` predicate.

aschackmull commented on java/ql/lib/semmle/code/java/ControlFlowGraph.qll:
```diff
@@ -434,6 +436,71 @@ private module ControlFlowGraphImpl {
+   * that any default case comes after the last pattern case.
+   */
+  private SwitchCase getASuccessorSwitchCase(PatternCase pred) {
+    result.getParent() = pred.getParent() and
```
This is needlessly quadratic, but I guess we'll have to hit at least thousands of `PatternCase`s in a single switch before this starts to bite, so I'm ok leaving it as-is for now.

smowton replied:
Explain how? I thought by only computing it for PatternCase and having each result only appear once I would dodge that, but I guess you mean the intermediate result will be quadratic until the limit on the max index (that of the next pattern case) is applied?

aschackmull replied:
> the intermediate result will be quadratic until the limit on the max index (that of the next pattern case) is applied?

Yes. The constraints on index (both lower and upper bounds) can only be applied as filters once the product has been constructed. We would instead need to construct the index range and then do a two-column join on parent+index. E.g. something like this:
```
pragma[nomagic]
private predicate switchCaseIndex0(StmtParent switch, SwitchCase c, int i) {
  c.getParent() = switch and c.getIndex() = i
}

private int lastIndex(StmtParent switch) { result = max(int i | switchCaseIndex0(switch, _, i) }

pragma[inline_late]
bindingset[switch, i]
private predicate switchCaseIndex(StmtParent switch, SwitchCase c, int i) {
  switchCaseIndex0(switch, c, i)
}

private SwitchCase getASuccessorSwitchCase(PatternCase pred) {
  // Note we do include `case null, default` (as well as plain old `default`) here.
  not result.(ConstCase).getValue(_) instanceof NullLiteral and
  exists(int end |
    if exists(getNextPatternCase(pred))
    then end = getNextPatternCase(pred).getIndex()
    else end = lastIndex(pred.getParent())
  |
    switchCaseIndex(pred.getParent(), result, [pred.getIndex() + 1 .. end])
  )
}
```
Alternatively, use a step relation:
```
private SwitchCase nextCase(SwitchCase c) {
  exists(StmtParent switch, int i | c = getCase(switch, i) and result = getCase(switch, i + 1))
}

private SwitchCase nextCaseFromNonPatternCase(SwitchCase c) {
  result = nextCase(c) and not c instanceof PatternCase
}

private  SwitchCase getASuccessorSwitchCase(PatternCase pred) {
  result = nextCaseFromNonPatternCase*(nextCase(pred)) and
   // Note we do include `case null, default` (as well as plain old `default`) here.
  not result.(ConstCase).getValue(_) instanceof NullLiteral
}
```
The latter fix is simpler, but depends slightly on the optimiser to do something reasonable with the TC.

aschackmull replied:
The latter could also be expanded to an explicit recursion, which would guarantee the desired asymptotic behaviour, but then the null case becomes slightly awkward - unless we could possible push its exclusion into the rank in `getCase` - depends on the needs of the potential other uses of `getCase`, I haven't checked that.

smowton replied:
Gone with your first solution, only with `switchCaseIndex0` turned into a member `isNthCaseOf` in the spirit of `isNthChildOf`.

smowton commented on java/ql/test/library-tests/guards12/guard.expected:
```diff
@@ -6,3 +6,5 @@
 | Test.java:11:7:11:17 | case ... | Test.java:9:13:9:13 | s | Test.java:11:12:11:14 | "c" | true | true | Test.java:11:7:11:17 | case ... |
 | Test.java:12:7:12:17 | case ... | Test.java:9:13:9:13 | s | Test.java:12:12:12:14 | "d" | true | false | Test.java:13:7:13:16 | default |
 | Test.java:12:7:12:17 | case ... | Test.java:9:13:9:13 | s | Test.java:12:12:12:14 | "d" | true | true | Test.java:12:7:12:17 | case ... |
+| Test.java:17:7:17:17 | case ... | Test.java:15:13:15:13 | s | Test.java:17:12:17:14 | "e" | true | false | Test.java:18:7:18:16 | default |
```
That's because `s.length()` isn't a VarAccess. Twiddled the test to get this to show up.

aschackmull replied:
Ah, didn't notice the type constraint in `guard.ql`.

smowton commented:
@aschackmull all today's comments addressed, starting from commit "Cast back to Object in advance of returning"

aschackmull commented on java/ql/lib/semmle/code/java/PrettyPrintAst.qll:
```diff
@@ -742,11 +745,19 @@ private class PpSwitchStmt extends PpAst, SwitchStmt {
-    i = 0 and result = "default" and this instanceof DefaultCase
+    i = 0 and result = "default" and isNonNullDefaultCase(this)
+    or
+    i = 0 and result = "case " and not isNonNullDefaultCase(this)
```
Right, the problem here is that `NullDefaultCase` isn't a `ConstCase`, so we don't get the `null` or the `", "` printed. Let's just handle the `NullDefaultCase` with a `i = 0 and result = "case null, default"`. Then we can also delete the `getCaseDefaultOffset()`.

aschackmull commented on java/ql/test/library-tests/prettyprint/pp.expected:
```diff
@@ -0,0 +1,80 @@
+| Test.java:1:14:1:17 | Test | 61 |     var a = switch (o) { |
+| Test.java:1:14:1:17 | Test | 62 |       case String s: |
+| Test.java:1:14:1:17 | Test | 63 |       yield 1; |
+| Test.java:1:14:1:17 | Test | 64 |       case R(S(x), y): |
```
We ought to print some types for `x` and `y`.

aschackmull commented on java/ql/lib/semmle/code/java/ControlFlowGraph.qll:
```diff
@@ -847,7 +919,7 @@ private module ControlFlowGraphImpl {
+      not completion instanceof NormalOrBooleanCompletion
       or
       // if the last case completes normally, then so does the switch
       last(switch.getStmt(strictcount(switch.getAStmt()) - 1), last, NormalCompletion()) and
```
In the lines below here (which I can't comment on since they're too far from the diff), which read
```
      // if no default case exists, then normal completion of the expression may terminate the switch
      not exists(switch.getDefaultCase()) and
      last(switch.getExpr(), last, completion) and
      completion = NormalCompletion()
```
we'll need to add `not exists(switch.getAPatternCase()), since we're doing sequential testing of the pattern cases and therefore shouldn't include the "nothing matched" edge.

smowton replied:
In fact the JLS requires exhaustiveness if either null or pattern cases are involved (i.e., we're sure this isn't a grandfathered pre-Java-21 switch block), so we know such an edge is impossible if the program compiles.

aschackmull commented on java/ql/lib/semmle/code/java/ControlFlowGraph.qll:
```diff
@@ -867,24 +939,50 @@ private module ControlFlowGraphImpl {
+    // to have explicit `yield` statements.
+    exists(SwitchExpr parent, SwitchCase case |
+      case = n and
+      case = parent.getACase() and
```
Actually, maybe we could simplify these two cases as follows:
```
    exists(Completion caseCompletion |
      last(n.(SwitchCase).getRuleStatementOrExpressionStatement(), last, caseCompletion) and
      if caseCompletion instanceof NormalOrBooleanCompletion
      then
        n.getParent() instanceof SwitchStmt and completion = anonymousBreakCompletion()
        or
        n.getParent() instanceof SwitchExpr and completion = YieldCompletion(caseCompletion)
      else completion = caseCompletion
    )
```
I _think_ this amounts to the same thing, right? It's considerably shorter.

aschackmull commented on java/ql/lib/semmle/code/java/ControlFlowGraph.qll:
```diff
@@ -867,24 +939,50 @@ private module ControlFlowGraphImpl {
+      )
+    )
+    or
+    // The normal last node in a non-rule pattern case is the last of its variable declaration(s),
+    // or the successful matching of its guard if it has one.
+    // Note that either rule or non-rule pattern cases can end with pattern match failure, whereupon
+    // they branch to the next candidate pattern. This is accounted for in the `succ` relation.
+    exists(PatternCase pc | n = pc |
+      (
+        if exists(pc.getGuard())
+        then last(pc.getGuard(), last, BooleanCompletion(true, _))
+        else last(pc.getPattern(), last, NormalCompletion())
+      ) and
+      not pc.isRule() and
+      completion = NormalCompletion()
```
I'd like to rework this, since the CFG design intends a clear split of responsibilities: `first` deals with half-edges entering an AST node, `last` deals with half-edges leaving an AST-node, and `succ` deals with those edges where both endpoints are within a given AST node. Thus enabling a nice structural recursion. The comment indicates that this split of responsibilities is broken here.
```suggestion
    exists(PatternCase pc | n = pc |
      last = n and completion = basicBooleanCompletion(false)
      or
      last(pc.getPattern(), last, NormalCompletion()) and
      not pc.isRule() and
      not exists(pc.getGuard()) and
      completion = basicBooleanCompletion(true)
      or
      last(pc.getGuard(), last, completion) and
      (
        completion = BooleanCompletion(true, _) and
        not pc.isRule()
        or
        completion = BooleanCompletion(false, _)
      )
```

aschackmull replied:
The suggested change should cover the half-edges we need in the `last` relation. It probably need a sprinkle of comments.

aschackmull commented on java/ql/lib/semmle/code/java/ControlFlowGraph.qll:
```diff
@@ -1201,10 +1299,14 @@ private module ControlFlowGraphImpl {
+      // (they always break or yield), and the case of pattern matching failure branching to the
+      // next case is specially handled in the `PatternCase` logic below.
       exists(int i |
         last(switch.getStmt(i), n, completion) and result = first(switch.getStmt(i + 1))
```
This case should now also handle `completion = BooleanCompletion(true, _)` when `switch.getStmt(i) instanceof PatternCase`. Same goes for the identical code below for `SwitchExpr`.

aschackmull replied:
We should also add the false-edges going from `last(switch.getStmt(i), n, completion)` to the set of possible next cases when `switch.getStmt(i) instanceof PatternCase`. These are edges that go between direct children of the switch and should therefore be accounted for in this section.

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -103,13 +103,28 @@ class Guard extends ExprParent {
+    // Return the closest pattern case statement before this one, including this one.
+    result =
+      max(int i, PatternCase c |
+        c = this.(SwitchCase).getSiblingCase(i) and i <= this.(SwitchCase).getCaseIndex()
```
As noted at the definition of `getSiblingCase`, this is quadratic, and we'll want to implement this differently.

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -137,13 +152,16 @@ class Guard extends ExprParent {
-      pred = sc.getControlFlowNode().getAPredecessor() and
-      pred.(Expr).getParent*() = sc.getSelectorExpr() and
-      bb1 = pred.getBasicBlock()
+      bb1 = sc.getControlFlowNode().getAPredecessor().getBasicBlock() and
+      // This is either the top of the switch block, or a preceding pattern case
+      // if one exists.
+      this.getBasicBlock() = bb1
```
The rewrite here fails to recognise a branch edge for
```
switch(b ? e1 : e2) {
case "a": ..
```

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -175,13 +193,74 @@ class Guard extends ExprParent {
+  }
+
+  /**
+   * Gets the record pattern this type test binds to, if any.
```
Not necessarily a record
```suggestion
   * Gets the pattern this type test binds to, if any.
```

aschackmull replied:
Or maybe just delete this predicate? I don't see any use of it.

aschackmull replied:
In fact, this leaves just the charpred and the `appliesTypeTest`, so maybe get rid of the class and just have an `isTypeTest` predicate directly on `Guard` similar to the `isEquality`?

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -175,13 +193,74 @@ class Guard extends ExprParent {
+   * the type test could mean a nested record or binding pattern didn't match but `testedExpr` is still
+   * of type `testedType`.)
+   */
+  predicate appliesTypeTest(Expr e, Type t, boolean testedBranch) {
```
The `testedBranch` column in this predicate is always either `[true,false]` or just `true`. Maybe we should change its name and represent those two possibilities as `true` and `false`, respectively. That way we don't duplicate all the `(Guard, Expr, Type)` tuples for which the test is an unrestricted pattern. Maybe a good name would be something like "faithful"?

smowton commented:
@aschackmull I am once again up to date with your comments

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/internal/SwitchCases.qll:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Gets the `i`th `SwitchCase` defined on `switch`, if one exists.
+ */
+SwitchCase getCase(StmtParent switch, int i) {
```
This is now a duplicate, so we can remove it and replace its uses with `SwitchCase.isNthCaseOf`.

aschackmull commented on java/ql/lib/semmle/code/java/ControlFlowGraph.qll:
```diff
@@ -867,24 +948,46 @@ private module ControlFlowGraphImpl {
+      last(pc.getGuard(), last, completion) and
+      (
+        completion = BooleanCompletion(false, _) or
+        not completion instanceof NormalOrBooleanCompletion
```
This line is superfluous. There's a catch-all propagation of these completions at the very top of the definition of `last`.

aschackmull commented on java/ql/lib/semmle/code/java/ControlFlowGraph.qll:
```diff
@@ -867,24 +948,46 @@ private module ControlFlowGraphImpl {
+    // A pattern case statement can complete:
+    // * On failure of its type test (boolean false)
+    // * On failure of its guard test if any (boolean false)
+    // * On any abrupt completion of its guard
```
So we can also remove this bullet.

aschackmull commented on java/ql/lib/semmle/code/java/ControlFlowGraph.qll:
```diff
@@ -1196,39 +1299,67 @@ private module ControlFlowGraphImpl {
-      // ...and then to one of the cases.
-      last(switch.getExpr(), n, completion) and result = first(switch.getACase())
+    // Switch statements and expressions
+    exists(StmtParent switch | switch instanceof SwitchStmt or switch instanceof SwitchExpr |
```
It's a bit confusing to see this here when it looks like it's implied by all the disjuncts. Maybe lets move it into `getSwitchStatement`, so that predicate actually becomes restricted to switches?

smowton replied:
Have introduced `SwitchBlock` to do this in many places -- will follow up once this is merged to deduplicate predicates from SwitchStmt and SwitchExpr and use this more widely.

smowton commented:
@aschackmull up to date again.

aschackmull commented on java/ql/lib/semmle/code/java/Statement.qll:
```diff
@@ -428,6 +457,21 @@ class SwitchCase extends Stmt, @case {
+   * Holds if this is the `n`th case of switch block `parent`.
+   */
+  pragma[nomagic]
+  predicate isNthCaseOf(StmtParent parent, int n) {
```
```suggestion
  predicate isNthCaseOf(SwitchBlock parent, int n) {
```

aschackmull commented on java/ql/lib/semmle/code/java/Expr.qll:
```diff
@@ -1612,18 +1670,86 @@ class LocalVariableDeclExpr extends Expr, @localvariabledeclexpr {
+  /**
+   * Gets the switch statement or expression whose pattern declares this identifier, if any.
+   */
+  StmtParent getAssociatedSwitch() {
```
```suggestion
  SwitchBlock getAssociatedSwitch() {
```

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -72,6 +73,51 @@ class ConditionBlock extends BasicBlock {
+ * involving pattern cases it can also mean flow from a previous pattern case's type
+ * test or guard failing and proceeding to then consider subsequent cases.
+ */
+private predicate isNonFallThroughPredecessor(SwitchCase sc, ControlFlowNode pred) {
```
Based on the qldoc, we should add `pred = sc.getControlFlowNode().getAPredecessor()`.

aschackmull commented on java/ql/lib/semmle/code/java/controlflow/Guards.qll:
```diff
@@ -139,10 +226,12 @@ class Guard extends ExprParent {
+      not sc instanceof PatternCase and
       branch = true and
       bb2.getFirstNode() = sc.getControlFlowNode() and
       pred = sc.getControlFlowNode().getAPredecessor() and
```
delete this line once it's been moved into `isNonFallThroughPredecessor`

smowton commented:
👍 @aschackmull all changes applied and rebased on main to run full CI

aschackmull reviewed (approved):
LGTM! Let's see what CI and DCA says, and if that looks good, then I think we're ready to merge.

smowton commented:
DCA is fine -- no timing changes, alert improvements on elasticsearch due to CFG fixes. Proceeding to QA.
